src/Features/Core/Portable/SignatureHelp/SignatureHelpService.cs

```diff
@@ -0,0 +1,126 @@
+            => _allProviders.Where(p => p.Metadata.Language == language)
+                .SelectAsArray(p => p.Value));
+    }
+
+    /// <summary>
+    /// Gets the <see cref="ISignatureHelpProvider"/> and <see cref="SignatureHelpItems"/> associated with
+    /// the position in the document.
+    /// </summary>
+    public Task<(ISignatureHelpProvider? provider, SignatureHelpItems? bestItems)> GetSignatureHelpAsync(
+        Document document,
+        int position,
+        SignatureHelpTriggerInfo triggerInfo,
+        SignatureHelpOptions options,
+        CancellationToken cancellationToken = default)
+    {
+        return GetSignatureHelpAsync(
+            GetProviders(document.Project.Language),
+            document,
+            position,
+            triggerInfo,
+            options,
+            cancellationToken);
+    }
+
+    /// <summary>
+    /// Gets the <see cref="ISignatureHelpProvider"/> and <see cref="SignatureHelpItems"/> associated with
+    /// the position in the document.
+    /// </summary>
+    public async Task<(ISignatureHelpProvider? provider, SignatureHelpItems? bestItems)> GetSignatureHelpAsync(
+        ImmutableArray<ISignatureHelpProvider> providers,
+        Document document,
+        int position,
+        SignatureHelpTriggerInfo triggerInfo,
+        SignatureHelpOptions options,
+        CancellationToken cancellationToken)
+    {
+        ISignatureHelpProvider? bestProvider = null;
+        SignatureHelpItems? bestItems = null;
+
+        // returns the first non-empty quick info found (based on provider order)
+        foreach (var provider in providers)
+        {
+            var items = await TryGetItemsAsync(document, position, triggerInfo, options, provider, cancellationToken).ConfigureAwait(false);
+            if (items is null)
+            {
+                continue;
+            }
+
+            // If another provider provides sig help items, then only take them if they
+            // start after the last batch of items.  i.e. we want the set of items that
+            // conceptually are closer to where the caret position is.  This way if you have:
+            //
+            //  Goo(new Bar($$
+            //
+            // Then invoking sig help will only show the items for "new Bar(" and not also
+            // the items for "Goo(..."
+            if (bestItems is not null && items.ApplicableSpan.Start < bestItems.ApplicableSpan.Start)
+            {
+                continue;
+            }
+
+            bestProvider = provider;
+            bestItems = items;
+        }
+
+        return (bestProvider, bestItems);
+    }
+
+    private static async Task<SignatureHelpItems?> TryGetItemsAsync(Document document, int position, SignatureHelpTriggerInfo triggerInfo, SignatureHelpOptions options, ISignatureHelpProvider provider, CancellationToken cancellationToken)
+    {
+        // We're calling into extensions, we need to make ourselves resilient
```

Comment:
Might be nice to wrap it in the extension manager stuff while we're making changes here, e.g. similar to https://github.com/dotnet/roslyn/blob/d89c824648207390f5be355a782048812ba5f91e/src/Features/Core/Portable/QuickInfo/QuickInfoServiceWithProviders.cs#L55